### ISSUE TITLE
My test results

### ISSUE THREAD

**Issue by Musixal** [EasyTier/EasyTier#139]
I recently used this project. Unfortunately, there are problems that I will explain in the future. The first problem is the low bandwidth it provides. In the best case, the speed cannot be more than 400 Mbps on plain TCP. Do you have a solution? thank you

**Comment by Musixal** [comment]
The second problem is when the number of users and connections increases, in this case the latency increases greatly and the process needs to be restarted.

The third problem is in the connection between a server with public IP and a client behind NAT. In this case, the connection is unstable and according to the monitoring I did, the pings increase drastically and sometimes the connection is interrupted.

All these tests have been done and confirmed by more than 5 people.

**Comment by Musixal** [comment]
If I want to make a comparison with a reverse proxy like Rathole (of course, I know Rathole is not a mesh, but I use it to connect a server to a client behind NAT)
Rathole has no problem in the connection above 20000 and keeps the tunnel stable without increasing the delay. But easytier does not provide secure tunnel in lower connections, maybe around 4000.
I think adding options such as setting keep alive time interval will help to improve its performance.

**Comment by KKRainbow** [comment]
> I recently used this project. Unfortunately, there are problems that I will explain in the future. The first problem is the low bandwidth it provides. In the best case, the speed cannot be more than 400 Mbps on plain TCP. Do you have a solution? thank you

could you provide these additional infomation:
1. what device & OS did you use for test
2. what args is used to launching easytier on your test nodes.
3. what tools and what commands did you used to do this test. so i can reproduce your test result.

I've done some perf test before and looks good, maybe recent code changes lead to degraded perf

**Comment by KKRainbow** [comment]
> The second problem is when the number of users and connections increases, in this case the latency increases greatly and the process needs to be restarted.

are these users are connected to same virtual net (which means network_name & secret is all same)。

currently all peers in one virtual net forms a full-connected network ( 2^n connection），and in this version the keepalive is very aggressive.

i have plan to optimize this in recent, with on-demand p2p and more smart keepalive strategy

> The third problem is in the connection between a server with public IP and a client behind NAT. In this case, the connection is unstable and according to the monitoring I did, the pings increase drastically and sometimes the connection is interrupted.

i'll try this test. thanks a lot for your reporting

**Comment by KKRainbow** [comment]
> If I want to make a comparison with a reverse proxy like Rathole (of course, I know Rathole is not a mesh, but I use it to connect a server to a client behind NAT) Rathole has no problem in the connection above 20000 and keeps the tunnel stable without increasing the delay. But easytier does not provide secure tunnel in lower connections, maybe around 4000. I think adding options such as setting keep alive time interval will help to improve its performance.

I'll try this too. I have not test for the large number of connections.

I'll work on these problems and hoping resolve these all in next release.

**Comment by Musixal** [comment]
> > I recently used this project. Unfortunately, there are problems that I will explain in the future. The first problem is the low bandwidth it provides. In the best case, the speed cannot be more than 400 Mbps on plain TCP. Do you have a solution? thank you
> 
> could you provide these additional infomation:
> 
> 1. what device & OS did you use for test
> 2. what args is used to launching easytier on your test nodes.
> 3. what tools and what commands did you used to do this test. so i can reproduce your test result.
> 
> I've done some perf test before and looks good, maybe recent code changes lead to degraded perf

All tests are taken on Ubuntu 22.04 with Xanmod kernel with 1 gig port. The results with the stock kernel are very low at 150Mb/s. It has been tested in all possible modes, both tcp and udp, and the maximum bandwidth of 400 Mb/s was obtained. Both IPv4 and IPv6 had the same results.

**Comment by Musixal** [comment]
> > The second problem is when the number of users and connections increases, in this case the latency increases greatly and the process needs to be restarted.
> 
> are these users are connected to same virtual net (which means network_name & secret is all same)。
> 
> currently all peers in one virtual net forms a full-connected network ( 2^n connection），and in this version the keepalive is very aggressive.
> 
> i have plan to optimize this in recent, with on-demand p2p and more smart keepalive strategy
> 
> > The third problem is in the connection between a server with public IP and a client behind NAT. In this case, the connection is unstable and according to the monitoring I did, the pings increase drastically and sometimes the connection is interrupted.
> 
> i'll try this test. thanks a lot for your reporting

Yes, all clients connected to the same virtual net with same secret and network name.

**Comment by Musixal** [comment]
> > If I want to make a comparison with a reverse proxy like Rathole (of course, I know Rathole is not a mesh, but I use it to connect a server to a client behind NAT) Rathole has no problem in the connection above 20000 and keeps the tunnel stable without increasing the delay. But easytier does not provide secure tunnel in lower connections, maybe around 4000. I think adding options such as setting keep alive time interval will help to improve its performance.
> 
> I'll try this too. I have not test for the large number of connections.
> 
> I'll work on these problems and hoping resolve these all in next release.

Maybe doing a benchmark test with Vegeta Attack is a good option to go.

**Comment by Musixal** [comment]
some results for iperf3 and monitoring attached. 
![iperf3](https://github.com/EasyTier/EasyTier/assets/124933548/b4b8bdef-848e-42dc-ab10-0a16bdf719f3)
![ping1](https://github.com/EasyTier/EasyTier/assets/124933548/9c8c82ce-cb82-4c8e-809c-27f27ba8a4e6)
![ping2](https://github.com/EasyTier/EasyTier/assets/124933548/40c6f4ad-266e-4fdb-9256-ee30ed3f0943)

**Comment by KKRainbow** [comment]
thanks for your tests and report. I'll try to resolve the stability & perfomance problems in this issue. https://github.com/EasyTier/EasyTier/issues/140

I'll also link related items to this issue.

**Comment by KKRainbow** [comment]
> But easytier does not provide secure tunnel in lower connections, maybe around 4000. I think adding options such as setting keep alive time interval will help to improve its performance.

how is the connection counted. all p2p connections in one virtual network ?

**Comment by Musixal** [comment]
> > But easytier does not provide secure tunnel in lower connections, maybe around 4000. I think adding options such as setting keep alive time interval will help to improve its performance.
> 
> how is the connection counted. all p2p connections in one virtual network ?
> 
> as for only 254 peer in one virtual network, for one server max 254 conn would be established.

For clarification, the internal network has been used as a secure tunnel, and xray core has managed the requests. For example, 100 users are connected to server A and tunneled to server B through the internal network created by easytier. In this scenario where the number of connections is high and easytier has not been stable and reliable.

**Comment by KKRainbow** [comment]
can i contact you with telegram.

my username: @kkrainbow_1
